New UX component Typed (#110)
symfony/ux: src/Typed/Resources/doc/index.rst

.. image:: Animation.gif
  :alt: Typed showing messages

Installation
------------

Before you start, make sure you have `Symfony UX configured in your app`_.

Then install the bundle using Composer and Symfony Flex:

.. code-block:: terminal

    $ composer require symfony/ux-typed

    # Don't forget to install the JavaScript dependencies as well and compile
    $ yarn install --force
    $ yarn encore dev

Also make sure you have at least version 3.0 of
`@symfony/stimulus-bridge`_ in your ``package.json`` file.

Usage
-----

Typed works using a list of strings and will manage typing them on your page.
It comes with a lot of parameters to customize the way the strings are typed: speed, cursor, delays
and smart backspace are some incredible parameters you can use.

The main usage of Symfony UX Typed is to use its Stimulus controller to initialize Typed:

.. code-block:: twig

    <div>
        I created this UX component because
        <span {{ stimulus_controller('symfony/ux-typed/typed', {
            strings: ['I ❤ Symfony UX', 'Symfony UX is great', 'Symfony UX is easy']
        }) }}></span>
    </div>

> .. note::

Reviewer: ```suggestion

```

> That's it! Typed now shows the messages defined in the list passed as argument.

Reviewer: ```suggestion
That's it! Typed now shows the messages defined in the `strings` argument.
```

You can customize the way those messages are typed.
Parameters are exactly the same as for the `typed library`_

.. code-block:: twig

    <div>
        I created this UX component because
        <span {{ stimulus_controller('symfony/ux-typed/typed', {
            strings: ['I ❤ Symfony UX', 'Symfony UX is great', 'Symfony UX is easy'],
            smartBackspace: true,
            startDelay: 100,
            backSpeed: 20,
            backDelay: 100,
            loop: true,
            showCursor: true,
            cursorChar: '✨'
        }) }}></span>

>     </div>

Reviewer: I love how you included an animated gif for the readme... and then you show how to exactly re-recreate that in the code. Super easy to understand!

Author: Frankly, the Stimulus course on SymfonyCasts inspired me a lot. All the credit goes to you!

.. note::

> Extend the default behavior

Reviewer: ```suggestion
Extend the JavaScript Controller
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~

Symfony UX Typed allows you to extend its default behavior using a custom
Stimulus controller:

.. code-block:: javascript

    // assets/controllers/mytyped_controller.js

    import { Controller } from '@hotwired/stimulus';

    export default class extends Controller {
        connect() {
            this.element.addEventListener('typed:pre-connect', this._onPreConnect);
            this.element.addEventListener('typed:connect', this._onConnect);

>         }

Reviewer: It's annoying, but to avoid the `this` problem in their functions, we should probably add:

```js
initialize() {
    // fix the "this" variable for both methods
    this._onPreConnect = this._onPreConnect.bind(this);
    this._onConnect = this._onConnect.bind(this);
}
```

Author: Argh yes. On of the coolest issue. I never understood how to get out of this one...

        disconnect() {
            // You should always remove listeners when the controller is disconnected to avoid side-effects
            this.element.removeEventListener('typed:pre-connect', this._onConnect);
            this.element.removeEventListener('typed:connect', this._onPreConnect);
        }

        _onPreConnect(event) {
            // Typed has not been initialized - options can be changed

>             console.log(event.detail.options); // Options that will be used to initialize Typed

Reviewer: Maybe also show:

```js
event.details.options.onBegin = (typed) => {
    // ...
});
```

Author: OK. I added few lines
